src/scikit_build_core/cli/utils.py

```diff
@@ -0,0 +1,40 @@
+from __future__ import annotations
+
+import functools
+import pathlib
+from typing import TYPE_CHECKING
+
+import click
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+    from typing import Any, TypeVar
+
+    F = TypeVar("F", bound=Callable[..., Any])
+
+
+__all__: list[str] = ["_build_dir"]
+
+
+def _build_dir(func: F) -> F:
+    """Add build_dir click option"""
+
+    @click.option(
+        "--build-dir",
+        "-B",
+        type=click.Path(
+            exists=True,
+            file_okay=False,
+            dir_okay=True,
+            writable=True,
+            path_type=pathlib.Path,
+        ),
+        help="Path to cmake build directory",
+    )
+    @functools.wraps(func)
+    # TODO: Fix mypy checks here.
+    #  See upstream approach: https://github.com/pallets/click/blob/main/src/click/decorators.py
+    def wrapper(*args, **kwargs):  # type: ignore[no-untyped-def]
+        return func(*args, **kwargs)
+
+    return wrapper  # type: ignore[return-value]
```

Comment:
Getting this to be `mypy` compatible is a bit troublesome. I think it would need `ParamSpec` and concatenation with specific `kwargs`. Need to test out what is the appropriate interface for the `wrapper` part.

Reply:
Just don't wrap it. Decorators are just functions. `_build_dir = click.option(...)` is exactly what you want already.